File: cpp/678.valid_parenthesis_string.cpp
#include <string>

class Solution {
 public:
    bool checkValidString(std::string s) {
        int leftMin = 0;
        int leftMax = 0;

        for (auto ch : s) {
            if (ch == '(') {
                ++leftMin;
                ++leftMax;
            } else if (ch == ')') {
                --leftMin;
                --leftMax;
            } else {
                --leftMin;
                ++leftMax;
            }

            if (leftMax < 0) {
                return false;
            }

            if (leftMin < 0) {
                leftMin = 0;
            }
        }

        if (leftMin == 0) {
            return true;
        }
        return false;
    }
};

File: python/678.valid_parenthesis_string.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        leftMin = 0
        leftMax = 0

        for element in s:
            if element == "(":
                leftMin += 1
                leftMax += 1
            elif element == ")":
                leftMin -= 1
                leftMax -= 1
            else:
                leftMin -= 1
                leftMax += 1

            if leftMax < 0:
                return False
            if leftMin < 0:
                leftMin = 0

        if leftMin == 0:
            return True
        return False
